### DISCUSSION TITLE
Right way to use use pnpm monorepo with TS library

### DISCUSSION THREAD

**Discussion by humphd** [pnpm/pnpm#7549] [Q&A]
I have a TS pnpm monorepo, it works amazingly (thank you!).

I'm adding a package to the workspace, and I need to:

1. use it as a dependency in other workspaces
2. build it as a CJS/ESM node.js package and publish to the npm registry

I'm struggling to figure out the right way to do this, such that I can get it to work in both configurations above.  If I define my `exports`, `main`, and `module` to point to the built files in `dist/`, it works for `npm install` users, but breaks in my monorepo.  If I set them to use files in `src/`, that won't work for JS users.

I know this has to be a solved problem, but I'm struggling to figure it out, and would value a pointer to an existing repo or example to follow.

Right now I'm using `unbuild`, but I could use any build tool for this.  I feel like the [publishConfig](https://pnpm.io/package_json#publishconfig) might provide a way out of this, where by default I can use `src/` files, and for published versions I can switch to `dist/`.  But I then need a build tool that is aware of these (unbuild uses the layout in `package.json` to configure things).

Thanks for any pointers.  Long time pnpm fan, first time caller!

**Comment by humphd** [comment]
I filed https://github.com/unjs/unbuild/issues/362 to see if unbuild could use the `publishConfig` to override inferred build config.  If people have a better solution, please educate me.

**Comment by dionysiusmarquis** [comment]
I'm running into the same scenario: I need `src` as import for all other apps in the monorepo and `dist` after publishing. Meaning `pnpm deploy` should keep `src` and `pnpm publish` should keep `dist`. Unfortunately `files` is not part of `publishConfig`.
